### PR TITLE
feat: restricted strategy list

### PR DESCRIPTION
**Pull request by 0xClandestine**

### CODE REVIEW

File: src/contracts/core/AllocationManager.sol [hunk 1]
```diff
@@ -53,13 +53,19 @@ contract AllocationManager is
...
         SlashingParams calldata params
     ) external onlyWhenNotPaused(PAUSED_OPERATOR_SLASHING) {
+        // Assert that the amount to slash is within bounds of 1 and WAD (1e18).
         require(0 < params.wadToSlash && params.wadToSlash <= WAD, InvalidWadToSlash());
 
-        // Check that the operator is registered and slashable
         OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: params.operatorSetId});
-        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
+
+        // Assert that the provided strategies are all whitelisted for the given operator set.
```
Comment (0xClandestine) [inline src/contracts/core/AllocationManager.sol:61]: Not a fan of repeat external calls to avsd, hoping we can merge these contracts.

File: src/contracts/core/AVSDirectory.sol [hunk 2]
```diff
@@ -413,6 +413,24 @@ contract AVSDirectory is
...
+        return _operatorSetStrategies[_encodeOperatorSet(operatorSet)].contains(address(strategy));
+    }
+
+    /// @inheritdoc IAVSDirectory
+    function isOperatorSetStrategyBatch(
+        OperatorSet calldata operatorSet,
+        IStrategy[] calldata strategies
+    ) external view returns (bool) {
+        for (uint256 i; i < strategies.length; ++i) {
+            if (!_operatorSetStrategies[_encodeOperatorSet(operatorSet)].contains(address(strategies[i]))) {
```
Comment (ypatil12) [inline src/contracts/core/AVSDirectory.sol:427]: Nit: can we save gas by not encoding every time?

File: src/contracts/core/AllocationManager.sol (hunk 1 above)
Comment (ypatil12) [reply src/contracts/core/AllocationManager.sol:61]: It's possible we can merge the _new_ AVSD functionality (ie. opset functionality) into the ALM. Unsure about merging ALM into AVSD

File: src/contracts/core/AllocationManager.sol [hunk 3]
```diff
@@ -53,13 +53,19 @@ contract AllocationManager is
...
+        require(avsDirectory.isOperatorSetStrategyBatch(operatorSet, params.strategies), InvalidStrategy());
+        // Asser that the operator is slashable for the given operator set.
         require(avsDirectory.isOperatorSlashable(params.operator, operatorSet), InvalidOperator());
 
+        // Check that the operator is registered and slashable
+        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
```
Comment (ypatil12) [inline src/contracts/core/AllocationManager.sol:67]: Nit: comment isn't doing this check? I think you mean to put above `require(avsDirectory.isOperatorSlashable(params.operator, operatorSet), InvalidOperator());`

File: src/contracts/core/AllocationManager.sol [hunk 4]
```diff
@@ -164,6 +170,11 @@ contract AllocationManager is
                     // Add the operatorSet to the deallocation queue
                     deallocationQueue[msg.sender][allocation.strategy].pushBack(operatorSetKey);
                 } else if (info.pendingDiff > 0) {
+                    require(
+                        avsDirectory.isOperatorSetStrategy(allocation.operatorSets[j], allocation.strategy),
```
Comment (ypatil12) [inline src/contracts/core/AllocationManager.sol:174]: The net effect here is that you should ONLY be able to allocate to strategies that are whitelisted but can deallocate regardless. I think this is great. Can we add this assumption to the interface?

File: src/contracts/interfaces/IAVSDirectory.sol [hunk 5]
```diff
@@ -323,6 +323,15 @@ interface IAVSDirectory is IAVSDirectoryEvents, IAVSDirectoryErrors, ISignatureU
         OperatorSet memory operatorSet
     ) external view returns (IStrategy[] memory strategies);
 
+    /// @notice Returns whether or not a strategy is allocatable/slashable for an operator set.
+    function isOperatorSetStrategy(OperatorSet calldata operatorSet, IStrategy strategy) external view returns (bool);
+
+    /// @notice Returns whether or not a list of strategies are all allocatable/slashable for an operator set.
```
Comment (ypatil12) [inline src/contracts/interfaces/IAVSDirectory.sol:329]: Nit: should we natspec the params?

File: src/test/unit/AllocationManagerUnit.t.sol [hunk 6]
```diff
@@ -111,6 +111,10 @@ contract AllocationManagerUnitTests is EigenLayerUnitTestSetup, IAllocationManag
         // Set operatorSet to being valid
         avsDirectoryMock.setIsOperatorSetBatch(operatorSets, true);
 
+        // Set strategy to being valid for operator set
```
Comment (ypatil12) [inline src/test/unit/AllocationManagerUnit.t.sol:114]: Can we add three tests: 

1. Revert when you try to slash for a strategy that does not exist. The operator should have allocated to the strategy, but then the AVS removes the strategy from their list of strategies we should revert the slash
2. Revert when the operator tries to allocate for a strategy not added to the operatorSet
3. Validate that we can deallocate, even if the strategy is not part of the operator set

File: src/contracts/core/AVSDirectory.sol (hunk 2 above)
Comment (ypatil12) [inline src/contracts/core/AVSDirectory.sol:427]: Can we add tests to the AVSD to validate this is calculated correctly? Just sanity checking the output from those fuzz tests

**Review by ypatil12 — COMMENTED** [review]
Mainly want some more test coverage

File: src/contracts/core/AVSDirectory.sol (hunk 2 above)
Comment (0xClandestine) [reply src/contracts/core/AVSDirectory.sol:427]: noted, will fix if we don't move new avsd logic.

File: src/contracts/core/AllocationManager.sol (hunk 4 above)
Comment (0xClandestine) [reply src/contracts/core/AllocationManager.sol:174]: You mean document it?

Comment (ypatil12) [reply src/contracts/core/AllocationManager.sol:173]: Ya in the interface comments

File: src/test/unit/AllocationManagerUnit.t.sol (hunk 6 above)
Comment (ypatil12) [reply src/test/unit/AllocationManagerUnit.t.sol:114]: I think we're still missing these?

**Review by ypatil12 — APPROVED** [review]

File: src/contracts/core/AllocationManager.sol [hunk 7]
```diff
@@ -53,13 +53,18 @@ contract AllocationManager is
...
     ) external onlyWhenNotPaused(PAUSED_OPERATOR_SLASHING) {
+        // Assert that the amount to slash is within bounds of 1 and WAD (1e18).
         require(0 < params.wadToSlash && params.wadToSlash <= WAD, InvalidWadToSlash());
 
-        // Check that the operator is registered and slashable
         OperatorSet memory operatorSet = OperatorSet({avs: msg.sender, operatorSetId: params.operatorSetId});
-        bytes32 operatorSetKey = _encodeOperatorSet(operatorSet);
+
+        // Assert that the provided strategies are all whitelisted for the given operator set.
+        require(avsDirectory.isOperatorSetStrategyBatch(operatorSet, params.strategies), InvalidStrategy());
```
Comment (wadealexc) [inline src/contracts/core/AllocationManager.sol:62]: Can we not have this method accept an operator set as an argument and fetch strategies from the AVSD? That seems far preferable to passing in an array of strategies here.

File: src/contracts/core/AllocationManager.sol [hunk 8]
```diff
@@ -53,13 +53,18 @@ contract AllocationManager is
...
+        // Asser that the operator is slashable for the given operator set.
```
Comment (wadealexc) [inline src/contracts/core/AllocationManager.sol:63]: `Asser` -> `Assert`

File: src/contracts/core/AllocationManager.sol (hunk 7 above)
Comment (0xClandestine) [reply src/contracts/core/AllocationManager.sol:62]: FIxed in #855, but the fix introduces a via-ir compile error. Thinking we wait until #845 is merged which should reduce code complexity.

**Pull request closed without merging**